enhancements/oc/cli-manager.md

```diff
@@ -0,0 +1,334 @@
+plugins via images, because this will enable offering plugins offline through an existing image registry.
+* Plugin owners must be able to easily distribute their binaries
+* Allow cluster-admins to control which plugins are offered to users
+  * This proposal is not concerned with _which_ plugins will be managed, as that is decided by cluster-admins
+* Controller for registering CRs with an API for listing, extracting and downloading plugins
+  * An API that generates an index that `krew` can consume via its custom index feature
+  * The index and binaries will served from a single route and service within the cluster
+* The `krew` client package will be vendored into `oc` for usage as `oc krew install <plugin>`
+
+### Non-Goals
+
+* The Controller will not build or package plugins
+* No recommending/limiting which plugins are served by the API
+* `krew` will not create or update the Plugin CRs, those will be managed by cluster-admins
+
+## Proposal
+
+Each component wishing to provide customers with their plugins will build and publish images via a trusted image registry
+and create a Plugin CR to provide an image name and the file path within that image for the binaries.
+Clients (i.e. `krew`) will read the index and download binaries from the Controller. The Controller is responsible for building the index from CRs, for pulling the images,
+extracting the binaries, and serving them to clients.
+
+* `krew` and `krew` plugins are upstream projects that Kubernetes users are already familiar with
+* A `krew`-compatible custom index can provide available plugins for a cluster in disconnected environments
+* The index will be served by a Controller with its contents managed by cluster-admins via CRs
+* The binaries will also be served by the Controller that will pull images from a trusted image registry and extract the binaries
+* With `krew index add https://someother-third-party-index` we won't limit users from adding their own index with whatever plugins they want
+* There are two possible approaches for this new controller:
+  1. Controller should be installed by default:
+  * Pros for being installed by default:
+    * Available day one for all clusters
+    * All binaries can be served using single unified mechanism
+    * No user setup required to use in disconnected environments
+    * Replace existing, temporary download mechanism
+  * Cons:
+    * Additional controller running by default
+  2. Controller being optional:
+  * Pros for being optionally installed:
+    * Less resource utilization
+  * Cons:
+    * Another operator to manage and mirror for disconnected environments
+    * Separate mechanism for built-in and third-party binaries
+    * Missing plugin CRs for optional elements installed before controller is installed
+    * Won't be able to replace existing, temporary system used by console which is serving some binaries
```

Comment:
not clear why it couldn't replace this also.

we don't necessarily need to solve this now, but long term i would still expect this component to be a candidate to move to olm, so we'll need to discuss it then

Reply:
If the controller is optional, then the console wouldn't be able to rely on it being present, so the existing mechanism would need to remain.

Reply:
> If the controller is optional, then the console wouldn't be able to rely on it being present, so the existing mechanism would need to remain.

no, if it's optional the console would have to be able to handle it not being present and tell the admin "hey if you want cli download functionality, please install/enable this optional feature/capability"